src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -801,6 +801,11 @@ void GcsActorManager::DestroyActor(const ActorID &actor_id,
         created_actors_.erase(node_it);
       }
     } else {
+      if (!actor->GetAddress().worker_id().empty()) {
+        // The actor is in phase of creating, so we need to notify the core
+        // worker exit to avoid process and resource leak.
+        NotifyCoreWorkerToKillActor(actor, death_cause, force_kill);
```

Comment:
is this mainly due to race conditions? also after this fix, will there still be case the actor is leaked?

Reply:
It,s not a race condition, if not fixed, tests in PR will definitely fail.

Reply:
I think what we need is to think about what metadata & resources are related to each state. 

Dependency unready -> pending creation -> created -> restarting -> dead

Here, pending creation has technically two parts, scheduling and actor initialization, and actor initialization has more resources (processes started) to clean up. I assume we could have a similar issue when the state is restarting (but the mechanism we added here should handle this correctly). This PR adds code that implicitly distinguish actor initialization vs scheduling (by checking if the actor has an associated worker)

I think to make the code really safe, we need to have state <-> in memory metadata + resources associated with it from the code level (so we can simply clean up those associated resources when it is destroyed), but it might require big changes.

Reply:
Yes, I think this PR can not handle every scenarios. e.g.

The actor is marked as dead and flushed to external storage before the kill RPC send to the core worker, and then GCS restart, in this case, the worker process may leak too.

Reply:
Seems we need to holistically think about the entire state machine and the corresponding resource management (not in this PR).